File: orderbook/bybit_orderbook.py
# orderbook/bybit_orderbook.py
import requests
import pandas as pd

class Exchange:
    def __init__(self, name, api_url):
        self.name = name
        self.api_url = api_url

    def fetch_orderbook(self):
        raise NotImplementedError()

class Bybit(Exchange):
    def fetch_orderbook(self):
        response = requests.get(self.api_url + 'v5/market/orderbook', params={'category': 'spot', 'symbol': 'BTCUSDT', 'limit': 50})
        data = response.json()
        bids = pd.DataFrame(data['result']['b'], columns=['Price', 'Size'])  # Corrected
        bids['Side'] = 'buy'
        asks = pd.DataFrame(data['result']['a'], columns=['Price', 'Size'])  # Corrected
        asks['Side'] = 'sell'
        return pd.concat([bids, asks], ignore_index=True)


        
if __name__ == '__main__':
    bybit = Bybit('Bybit', 'https://api.bybit.com/')
    print(bybit.fetch_orderbook())

File: orderbook_C++_implementation/src/bybit_orderbook.cpp
// src/bybit_orderbook.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include "../lib/json.hpp"
#include <curl/curl.h>
#include "bybit_orderbook.h"

using json = nlohmann::json;
using namespace std;

Bybit::Bybit(string name, string api_url) : Exchange(name, api_url) {}

vector<vector<string>> Bybit::fetch_orderbook() {
    string response = http_get(api_url + "v5/market/orderbook?category=spot&symbol=BTCUSDT&limit=50");
    json data = json::parse(response);

    vector<vector<string>> bids, asks;
    for (auto& item : data["result"]["b"]) {
        bids.push_back({item[0].get<string>(), item[1].get<string>(), "buy"});
    }
    for (auto& item : data["result"]["a"]) {
        asks.push_back({item[0].get<string>(), item[1].get<string>(), "sell"});
    }
    bids.insert(bids.end(), asks.begin(), asks.end());
    return bids;
}
